src/plugins/shell/shell_impl.cpp

```diff
@@ -18,60 +18,86 @@ void ShellImpl::init()
         this);
 }
 
-void ShellImpl::deinit() {}
+void ShellImpl::deinit()
+{
+    _parent->unregister_all_mavlink_message_handlers(this);
+}
 
 void ShellImpl::enable() {}
 
 void ShellImpl::disable() {}
 
-ShellImpl::ShellImpl(System& system) : PluginImplBase(system)
+ShellImpl::ShellImpl(System& system) :
+    PluginImplBase(system),
+    _transfer_closed_future(_transfer_closed_promise.get_future())
 {
     _parent->register_plugin(this);
+    _transfer_closed_promise.set_value();
 }
 
 ShellImpl::~ShellImpl()
 {
     _parent->unregister_plugin(this);
 }
 
+bool ShellImpl::is_transfer_in_progress()
+{
+    if (!_transfer_closed_future.valid()) {
+        return true;
+    }
+    return _transfer_closed_future.wait_for(std::chrono::seconds(0)) != std::future_status::ready;
+}
+
 void ShellImpl::shell_command_response_async(Shell::result_callback_t& callback)
 {
-    std::lock_guard<std::mutex> lock_subscription(_result_subscription_mutex);
+    {
+        std::lock_guard<std::mutex> lock(_future_mutex);
+        if (is_transfer_in_progress()) {
+            return;
+        }
+    }
     _result_subscription = callback;
 }
 
+void ShellImpl::send_to_subscription(Shell::Result arg)
+{
+    if (_result_subscription) {
+        auto callback = _result_subscription;
+        _parent->call_user_callback([callback, arg]() { callback(arg); });
+    }
+}
+
 void ShellImpl::shell_command(const Shell::ShellMessage& shell_message)
 {
-    std::lock_guard<std::mutex> lock_subscription(_result_subscription_mutex);
-    if (!_parent->is_connected()) {
-        if (_result_subscription) {
-            auto callback = _result_subscription;
-            Shell::Result arg{Shell::Result::ResultCode::NO_SYSTEM, {}};
-            _parent->call_user_callback([callback, arg]() { callback(arg); });
+    {
+        std::lock_guard<std::mutex> lock(_future_mutex);
+        if (is_transfer_in_progress()) {
             return;
```

Comment:
It is not good to just `return` if current transfer not completed yet.
May be need to return some `bool` value? 
What do you think?

Reply:
Usually I would return a `Shell::Result` and in this case it would say `Shell::Result::BUSY`.

Reply:
[Done](https://github.com/mavlink/MAVSDK/pull/878/commits/d985378c1e4bb4ef93d9fa02d7f6352916d8158e#diff-f4c5701958a5b01cfeef470e26440de5R49)